### PR TITLE
OCM-2501: OCM-2512: Improve upgrade semantics

### PR DESCRIPTION
**Pull request by JohnStrunk**
- Error if the requested version set to a value < current cluster version
- Cancel pending upgrades if requested version is set to current version

### CODE REVIEW

**Comment by JohnStrunk** [comment]
/cc @nirarg
FYI @amalykhi

**Comment by JohnStrunk** [comment]
/retest
Looks like CI is broken... related to rename?

File: provider/cluster_rosa_classic_resource.go [hunk 1]
```diff
@@ -1326,47 +1340,55 @@ func (r *ClusterRosaClassicResource) upgradeClusterIfNeeded(ctx context.Context,
...
+	if currentVersion.GreaterThan(desiredVersion) {
 		tflog.Debug(ctx, "No cluster version upgrade needed.")
+		if requestedVersionChanged {
+			// User changed the version they want, but actual is higher. We
+			// don't support downgrades.
+			return fmt.Errorf("cluster version is already above the requested version")
+		}
 		return nil
 	}
+	cancelingUpgradeOnly := desiredVersion.Equal(currentVersion)
```
Comment (bardielle) [inline provider/cluster_rosa_classic_resource.go:1352]: correct me if I'm wrong, but it seems that if cancelingUpgradeOnly is true - you can return from that function

File: provider/cluster_rosa_classic_resource.go [hunk 2]
```diff
@@ -1326,47 +1340,55 @@ func (r *ClusterRosaClassicResource) upgradeClusterIfNeeded(ctx context.Context,
...
-	availableVersions, err := upgrade.GetAvailableUpgradeVersions(ctx, r.versionCollection, versionId)
-	if err != nil {
-		return fmt.Errorf("failed to get available upgrades: %v", err)
-	}
-	found := false
-	for _, v := range availableVersions {
-		sem, err := semver.NewVersion(v.RawID())
-		if err != nil {
-			return fmt.Errorf("failed to parse available upgrade version: %v", err)
+	if !cancelingUpgradeOnly {
```
Comment (bardielle) [inline provider/cluster_rosa_classic_resource.go:1354]: That function become to be complicated - Can you refactor that code and move it to another function?

File: provider/cluster_rosa_classic_resource.go [hunk 3]
```diff
@@ -1381,7 +1403,7 @@ func (r *ClusterRosaClassicResource) upgradeClusterIfNeeded(ctx context.Context,
 		return err
 	}
 
-	if !correctUpgradePending {
+	if !correctUpgradePending && !cancelingUpgradeOnly {
 		// Gate agreements are checked when the upgrade is scheduled, resulting
```
Comment (bardielle) [inline provider/cluster_rosa_classic_resource.go:1407]: Also I'm prefer to refactor that logic and move it to another function (lines 1407- 1460) if it possible

**Comment by JohnStrunk** [comment]
I think the flow of upgradeClusterIfNeeded is better now that the 2 if-clauses were factored out.
/cc @bardielle

**Comment by JohnStrunk** [comment]
/retest
maybe there will be quota today

**Pull request merged**